### ISSUE TITLE
Tool Material Modifier Ideas

### ISSUE THREAD

**Issue by IntegratedQuantum** [PixelGuys/Cubyz#172]
Material should have certain modifiers associated with them, similar to how it's done in Tinker's Construct.
However given that this is a pattern based system, it would also be possible to allow modifiers that only work if certain patterns are made.

If you got any ideas for modifiers, feel free to comment them below.

### Common/Simple modifiers
Name | Concept | Examples
-- | -- | --
binding | Adds durability when next to a junction of other materials | sticky materials, fibers
long(?) | Stronger when multiple items of this type are in a straight/diagonal line | sticks/rods
hard | Increases speed when at the tip of a tool | diamond
brittle | Decreases durability depending on the number of neighbors surrounding it | diamond, glass/ceramic
thorny | Increases damage when on the blade, damages the player when on the handle | cactus
poisonous | Poison damage (damages the player when used for the handle) | lead, radioactive materials
combustible | The tool looses durability when the player is on fire | wood
hot | Burns the player/enemies, interacts with combustible materials | hot stuff?
cold | Protects neighboring combustible blocks from durability damage, looses durability when next to a hot material | ice
soluble | The tool looses durability when the player is under water | some salts/crystals
far-reaching | increases range of the tool

### Rare/special modifiers
Name | Concept
-- | --
single-use | only one material with this modifier can be used per tool, used to prevent overpowered tools
one-hit | increases damage to 1000×, but tools breaks after one use, without options to repair it
unbreakable | tool cannot loose durability
self-repairing | tool slowly heals durability
explosive | triggers a small explosion on use

### Uncategorized
- The surrounding 3×3 field needs to be filled with mundane items to enable some other modifier
- The tool must contain x copies of this material to enable some other modifier
- Activates another modifier if the tool contains no single-use materials
- Activates another modifier if the tool only contains a small amount of items

**Comment by Quazae303** [comment]
rare modifier: non-newtonian: looses power over time when not used, but if used often, gains power.

**Comment by FeaturesOfTheSAS** [comment]
heavy: takes more time to swing, deals more damage and breaks blocks easier

**Comment by IntegratedQuantum** [comment]
Material density already impacts these properties. A separate modifier for heavy tools is not needed.

**Comment by Argmaster** [comment]
It could be of use to add a modifier that would make tool very effective against specific block type/tag. For example, lets assume you want to dig a big hole in a mountain of sandstone. You could just make a pickaxe that is effective against every stone type, but it just happens so the mountain is made solely of stone. In such case you could opt for a tool that is very effective against particular type of material, lets say twice as effective as reachable by other means, while also twice less effective against other materials. This kind of specialization encourages players to craft more and put more effort into what they craft.

**Comment by IntegratedQuantum** [comment]
So cool that I immediately implemented it.

**Comment by Argmaster** [comment]
`blunt` - as durability approaches 0, damage of tool approaches 1. This can be used to penalize using soft materials on tip.

**Comment by Argmaster** [comment]
I think that instead of just increasing damage, `thorny` could inflict `bleeding` effect that would stack up to N times and deal small amount of damage per second, like 0.01, 0.02 with 2 stacks etc. Stacks would wear off after N minutes. Compared to fire, it does not cancel with frost and it can not be removed immediately. Player should not receive any special effects (like shaking of screen) when damage is inflicted.

**Comment by Argmaster** [comment]
We could have modifier like `life steal` or something that would heal you for small portion of damage delt. Could be provided by `ruby` gemstone.

**Comment by Argmaster** [comment]
We could have `energy steal` or something that would steal small portion of energy of your oponent (if they have energy) and give small amount of energy to you. Could be provided by `citrite` gemstone.

**Comment by Argmaster** [comment]
We could have `soul stealer` modifier that would cause blades to deal more damage more entities you killed with it. This is infinite scaling, so generally speaking should not appear at the same time as `unbreaking` or `self-repairing` to avoid infinite damage stacking. 

Overall I am not sold on having `self-repairing` and `unbreaking` in current form, in my opinion they will only cause players to craft less and need less resources. IMO if materials are fairly easy to obtain (as they can just be found in caves) it doesn't make sense to stay with same pickaxe infinitely. This is a bit different in games like Minecraft where creating a top pickaxe is quite expensive and doing this over and over is simply boring. In Cubyz with horizontal material balance I don't think we will have similar problem, as making even best pickaxe still only requires only few items, not hours of farming XP for enchants.

**Comment by IntegratedQuantum** [comment]
Those are good ideas, I added them to the list.
About unbreaking/self-repairing, I think they make a lot of sense if they are a tradeoff, maybe it takes away the players energy or health. (or maybe it's just blunt, making it useless over time anyways :​P)

**Comment by careeoki** [comment]
An idea I had for copper was "Conductive" which would increase strength of metals in adjacent slots.

**Comment by IntegratedQuantum** [comment]
Hmm, that is interesting. Though I would honestly want conductive for a different thing, like e.g. if you connect your handle conductively to anything with the hot modifier it would burn you.

**Comment by Argmaster** [comment]
We could have `fission` modifier for uranium that would behave like vein miner, basically breaking all the ores that directly touch. This would also require breaking blocks those ores are placed on, unless item unstucking would be able to reliably figure out how to push those items towards player. We could have range limit based on amount of uranium in tool, 1 block radius for one ingot, 2 for 2 etc.

**Comment by Argmaster** [comment]
Suggested during playtest:

- `electrocute` - it has a chance to electrocute enemies if used on blade. Electrocute chance and damage is boosted when used with metals, it is reduced otherwise. Copper has higher boost to chance and damage than other metals. Electrocute can be transferred between enemies, if they are close enough or in water. Distance threshold could scale with damage so it is higher for copper for exponential gain.

**Comment by Argmaster** [comment]
Suggested by @careeoki during playtest for #1300:

- `luck` - chance to do double damage to block/entity, to not lose durability, double drop chance for rare things (apples?) or generally to double drops.

**Comment by Argmaster** [comment]
`sticky` - When you mine a block with a tool that has sticky modifier it does not drop item, instead item is placed directly into your inventory.
Spawn rules

**Comment by Argmaster** [comment]
`wide swing` - deal same damage but in a horizontal cone, effectively allowing you to mine multiple blocks at once (requires #1261 or alike) - I am not sure how useful that would be for axes or pickaxes, but would definitely be useful for #1547

**Comment by vertigofilip** [comment]
Where is the up to date list of properties that was mentioned here?

**Comment by Argmaster** [comment]
Could you elaborate? What does up to date mean? List of implement - source code. List of planned - issue description. List of planned per material - see sub issues for #1269

**Comment by vertigofilip** [comment]
I am talking about https://github.com/PixelGuys/Cubyz/issues/172#issuecomment-2776165990

**Comment by Argmaster** [comment]
See issue description. This is the list of accepted modifier ideas.

**Comment by Argmaster** [comment]
Ruby could give powerful effect 2x of diamond, but reduce player max health 2x when held. When player changes slot hearts come back, but empty. This is similar to shaped glass effect from Risk Of Rain 2
Ideally effect would remain if tool is in inventory, but that may be hard to implement in generic way.

**Comment by ikabod-kee** [comment]
Except Shaped Glass is a lunar item that only comes off with a Cleansing Pool. Also imagine accidentally scrolling over the slot and losing half of your health.

**Comment by Argmaster** [comment]
> Except Shaped Glass is a lunar item that only comes off with a Cleansing Pool. Also imagine accidentally scrolling over the slot and losing half of your health.

Yep, intended interaction. Maybe it would be better as an effect from accessorily, easier to implement. As for the cleansing pool, no idea what you are talking about, I don't know ror2 that well, but I like the trade-off.

**Comment by careeoki** [comment]
> Also imagine accidentally scrolling over the slot and losing half of your health.




Maybe it could be a health drain effect (draining towards 50%) while you are holding it, like the gloves of running urgently in TF2.

**Comment by ikabod-kee** [comment]
More games need stuff from tf2. I'm being for real.

**Comment by careeoki** [comment]
Some misc ideas I had:

**Graceful**
Movement speed is increased while holding this tool

**Cumbersome** 
Movement speed is decreased while holding this tool

**Edible**
You can eat this tool, which restores some energy at the cost of durability

**Comment by codemob-dev** [comment]
> Edible
You can eat this tool, which restores some energy at the cost of durability

I don't know what to think about this

**Comment by careeoki** [comment]
I think it would be fun for candy cane tools.

**Comment by IntegratedQuantum** [comment]
I think stat boosts should be reserved to equipable items

**Comment by Bloxxel64** [comment]
That's true, but I think it would make sense if some applied to tools, mining efficiency, possibly like extra yield. I think stat boosts on tools make sense if they're stat boosts that make sense in context.

**Comment by IntegratedQuantum** [comment]
I disagree. I think it is fine, and maybe even desirable, to have a material that boosts speed when used in accessories and has no effect at all in tools.

**Comment by Crepestrom** [comment]
silk touch on a material

**Comment by Crepestrom** [comment]
On hit: builds fireball charges
fireball charges can be released as a ranged attack on left click

**Comment by Argmaster** [comment]
There is no need for silk touch in Cubyz.

**Comment by Bloxxel64** [comment]
Like I said, I think they should make sense in context. I don't think fireballs appearing make much sense.

**Comment by Crepestrom** [comment]
Randomizes your localization

**Comment by Argmaster** [comment]
So I've been thinking (it's bad when I think) how to make mining multiple blocks fun and we may do a little bit experimentation with multi block damage (yay another idea we can't implement now!)
Firstly, some time ago on a play session I suggested that flowers could cause the pickaxe deal damage in different patterns. While it was dismissed for flowers as they are more of a food/potion material, I still think we could figure out if eg. Gems could do that, like dealing damage in 3x3 cross shape, diamond shape, 2x2x2 volume and alike.
Secondly, explosive modified could be something along the lines of "every N hits triggers an explosion that breaks surrounding damaged blocks in MxM area". So, basically "choose your shape" option. Since there is a lot of parameters, there are plenty ways to tweak it. Shape choice doesn't require UI. For every 4th hit, a player could deal damage to 4 blocks and finish those for blocks. Blocks would have to be reasonably close together. 
Third option would be just the finisher, either aoe or not, to synergize with dynamite.
Those thingies I mentioned could be result of something like encasing with sulfur or alike.
God damn it sulfur would be useful.

**Comment by Argmaster** [comment]
There could be a bunch of accumulating counters, like **frostbite** that would debuff / damage player when the tool is held for too long so they have to change slots. Annoying, subtle suggestion to not use some material to make tools.

**Comment by Argmaster** [comment]
What could a mix of ice and magma do? That's kind of elemental boom, would deserve a fun interaction.

**Comment by Argmaster** [comment]
Maybe a brittle modifier could make the tool break when used against material that requires higher damage? That's punishing, but also easy to explain. Would have to be paired with some good positive modifier.

**Comment by IntegratedQuantum** [comment]
> choose your shape

Yeah, I've also been thinking about that, but instead of offering fixed shapes I think it could be more interesting as a "craft your shape" option, where material configuration (and not just whether or not it has a specific gem) determines the shape.
e.g. we could have primitive gems that produce tetris shapes and then materials that act on these shapes (e.g. to bring back your flowers: We could have a flower that rotates the shape of all neighboring gems by +90°, and further ones for the opposite direction, flipping and translation)

**Comment by ikabod-kee** [comment]
> thorny	Inflicts bleeding effect on the target, damages the player when on the handle	cactus

This should totally be called "serrated" :)

**Comment by calebstenberg** [comment]
Enveloping Durability - Replaces adjacent materials durability with its own.
Enveloping Hardness - Replaces adjacent materials hardness with its own.
I don't think there should be one for mass because it's too unrealistic.

**Comment by NovaFlag** [comment]
I feel like the "single use" modifier should be renamed to "pinnacle" or some similar word to more clearly signal to the player that that is one of the most powerful materials in the game and that they can only have one pinnacle per tool
